networkx/algorithms/euler.py

```diff
@@ -147,3 +158,203 @@ def eulerian_circuit(G, source=None):
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
+    -------
+    True, False
+
+    See Also
+    --------
+    is_eulerian()
+
+    Examples
+    --------
+    >>> G = nx.DiGraph([(1,2),(2,3)])
+    >>> nx.is_semieulerian(G)
+    True
+    """
+    is_directed = G.is_directed()
+
+    # Verify that graph is connected, short circuit
+    if is_directed and not nx.is_weakly_connected(G):
+        return False
+
+    # is undirected
+    if not is_directed and not nx.is_connected(G):
+        return False
+
+    # Not all vertex have even degree, check if exactly two vertex
+    # have odd degrees.  If yes, then there is an Euler path. If not,
+    # raise an error (no Euler path can be found)
+
+    # if the odd condition is not meet, raise an error.
+    start = _find_path_start(G)
+    if not start:
+        return False
+
+    return True
+
+
+def _find_path_start(G):
+    """Returns a suitable starting vertex for Eulerian path
+
+    The function also verifies that the graph contains a path. If no
+    path exists, returns ``None''.
+
+    """
+    is_directed = G.is_directed()
+
+    # list to check the odd degree condition
+    check_odd = []
+
+    if is_directed:
+        degree = G.in_degree
+        out_degree = G.out_degree
+    else:
+        degree = G.degree
+
+    # Verify if an Euler path can be found. Complexity O(|V|)
+    for v in G:
+        deg = degree(v)
+        # directed case
+        if is_directed:
+            outdeg = out_degree(v)
+            if deg != outdeg:
+                # if we have more than 2 odd nodes no Euler path exists
+                if len(check_odd) > 2:
+                    return False
+                # is odd and we append it.
+                check_odd.append(v)
+        # undirected case
+        else:
+            if deg % 2 != 0:
+                # if we have more than 2 odd nodes no Euler path exists
+                if len(check_odd) > 2:
+                    return False
+                # is odd and we append it.
+                check_odd.append(v)
+
+    if is_directed:
+
+        first = check_odd[0]
+        second = check_odd[1]
+        if G.out_degree(first) == G.in_degree(first) + 1 and \
+           G.in_degree(second) == G.out_degree(second) + 1:
+            start = second
+        elif G.out_degree(second) == G.in_degree(second) + 1 and \
+             G.in_degree(first) == G.out_degree(first) + 1:
+            start = first
+        else:
+            start = None
+
+    else:
+        start = check_odd[0]
+
+    return start
+
+
+def eulerian_path(G):
+    """Return a generator of the edges of an Eulerian path in ``G``.
+
+    Check if the graph ``G`` has an Eulerian path and return a
+    generator for the edges. If no path is available, raise an error.
+
+    Parameters
+    ----------
+    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
```

Comment:
This should remain a ```Returns``` section.
The function is a generator that returns an iterator (that's what a generator does in my understanding).
The iterator yields the results. The line below (281) should refer to ```edges``` as an iterator, not a generator.